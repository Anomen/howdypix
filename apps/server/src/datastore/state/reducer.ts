import { Reducer } from "redux";
import { uniqWith, isEqual } from "lodash";

type StateSearch = {
  lastUpdatedPhoto: Date | null;
  album: string;
  source: string;
};

type StateAlbum = {
  lastUpdatedPhoto: Date | null;
};

type StateSearchAlbumRelation = {
  searchId: number;
  albumId: number;
};

export type State = {
  searches: {
    [id: number]: StateSearch;
  };
  albums: {
    [id: number]: StateAlbum;
  };
  searchAlbumRelation: StateSearchAlbumRelation[];
};

export type Action =
  | {
      type: "RESET_STORE";
      data: State;
    }
  | {
      type: "UPDATE_SEARCH_LAST_UPDATED_PHOTO";
      searchId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_SEARCH";
      searchId: number;
      data: StateSearch;
    }
  | {
      type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO";
      albumId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_ALBUM";
      albumId: number;
      data: StateAlbum;
    }
  | {
      type: "STORE_NEW_SEARCH_ALBUM_RELATION";
      albumId: number;
      searchId: number;
    };

// ========================================================================
// Reducer
// ========================================================================
export const reducer: Reducer<State, Action> = (
  state = { searches: {}, albums: {}, searchAlbumRelation: [] },
  action
) => {
  // console.log(action);
  switch (action.type) {
    case "RESET_STORE":
      return action.data;

    case "UPDATE_SEARCH_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto =
        state.searches[action.searchId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          searches: {
            ...state.searches,
            [action.searchId]: {
              ...state.searches[action.searchId],
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "UPDATE_ALBUM_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto = state.albums[action.albumId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          albums: {
            ...state.albums,
            [action.albumId]: {
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "STORE_NEW_ALBUM":
      return {
        ...state,
        albums: {
          ...state.albums,
          [action.albumId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
      };

    case "STORE_NEW_SEARCH":
      return {
        ...state,
        searches: {
          ...state.searches,
          [action.searchId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
        searchAlbumRelation: uniqWith(
          [
            ...state.searchAlbumRelation,
            {
              searchId: action.searchId,
              albumId: action.data.album,
            },
          ],
          isEqual
        ),
      };

    case "STORE_NEW_SEARCH_ALBUM_RELATION":
      return {
        ...state,
        searchAlbumRelation: [
          ...state.searchAlbumRelation,
          ...(state.searchAlbumRelation.find(
            (relation) =>
              relation.albumId === action.albumId &&
              relation.searchId === action.searchId
          )
            ? [{ albumId: action.albumId, searchId: action.searchId }]
            : []),
        ],
      };

    default:
      return state;
  }
};
