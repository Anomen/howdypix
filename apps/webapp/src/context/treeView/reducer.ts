import { uniq, uniqBy } from "lodash";
import { QueryLazyOptions } from "@apollo/react-hooks";
import {
  GetTreeAlbums,
  GetTreeQuery,
  GetTreeQueryVariables,
  GetTreeSources,
} from "../../__generated__/schema-types";
import {
  AlbumWithNodeId,
  SourceWithNodeId,
  TreeItem,
  TreeItemWithParent,
} from "./types";

// ========================================================================
// Utility functions
// ========================================================================
const selectParents = (
  data: GetTreeQuery["getTree"]["albums"] | null | undefined,
  dir: string
): { [key: string]: boolean } => {
  let ret: { [key: string]: boolean } = {};

  if (!data) {
    ret[dir] = true;
  } else {
    data.forEach((d): void => {
      if (d.dir === dir) {
        ret[d.dir] = true;

        if (d.parentDir) {
          ret = {
            ...ret,
            ...selectParents(data, d.parentDir),
          };
        }
      }
    });
  }

  return ret;
};

const calculateExpandedNodeIds = (
  data: AlbumWithNodeId[],
  selectedItems: { [name: string]: boolean }
): string[] => {
  const ret: string[] = [];

  data.forEach((d): void => {
    if (d.dir && selectedItems[d.dir]) {
      ret.push(d.nodeId);
    }
    if (d.parentDir === "" && selectedItems[d.source]) {
      ret.push(d.source);
    }
  });

  return uniq(ret);
};

// ========================================================================
// Types
// ========================================================================
export type State = {
  fetchedAlbums: AlbumWithNodeId[];
  fetchedSources: SourceWithNodeId[];
  visibleLeaves: {
    [key: string]: boolean;
  };
  expandedNodeIds: string[];
};

export type Actions =
  | {
      type: "DATA_FETCHED";
      albums: GetTreeAlbums[];
      sources: GetTreeSources[];
      variables: {
        album: string;
        source: string;
      };
    }
  | {
      type: "DISPLAY_LEAF";
      album: string;
      source: string;
    }
  | {
      type: "UPDATE_EXPANDED_NODE_IDS";
    }
  | (TreeItem & {
      type: "EXPAND";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    })
  | (TreeItemWithParent & {
      type: "COLLAPSE";
    })
  | (TreeItemWithParent & {
      type: "TOGGLE";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    });

// ========================================================================
// Reducer
// ========================================================================
export const reducer = (state: State, action: Actions): State => {
  let newState: State = state;

  switch (action.type) {
    case "DATA_FETCHED":
      newState = {
        ...newState,
        fetchedSources: uniqBy(
          [...(newState.fetchedSources ?? []), ...action.sources],
          "name"
        ).map(
          (source): SourceWithNodeId => ({
            ...source,
            nodeId: source.name,
          })
        ),
        fetchedAlbums: uniqBy(
          [...(newState.fetchedAlbums ?? []), ...action.albums],
          "dir"
        ).map((album) => ({
          ...album,
          nodeId: album.source + album.dir,
        })),
      };

      return reducer(newState, {
        type: "DISPLAY_LEAF",
        source: action.variables.source,
        album: action.variables.album,
      });

    case "DISPLAY_LEAF":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== ""
              ? selectParents(newState.fetchedAlbums, action.album)
              : {}),
            [action.source]: true,
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "UPDATE_EXPANDED_NODE_IDS":
      return {
        ...newState,
        expandedNodeIds: calculateExpandedNodeIds(
          newState.fetchedAlbums,
          newState.visibleLeaves
        ),
      };

    case "EXPAND":
      if (action.album === null) {
        if (newState.visibleLeaves[action.source] === undefined) {
          action.fetchTree({
            variables: {
              album: "",
              source: action.source,
            },
          });
        }
      } else if (newState.visibleLeaves[action.album] === undefined) {
        action.fetchTree({
          variables: {
            album: action.album ?? "",
            source: action.source,
          },
        });
      }

      // different than original
      return reducer(
        reducer(newState, {
          type: "DISPLAY_LEAF",
          source: action.source,
          album: action.album ?? "",
        }),
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "COLLAPSE":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== null
              ? { [action.album]: false }
              : { [action.source]: false }),
            ...(action.parent ? { [action.parent]: true } : {}),
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "TOGGLE":
      if (action.album !== null) {
        if (newState.visibleLeaves[action.album]) {
          return reducer(newState, {
            ...action,
            type: "COLLAPSE",
          });
        }

        return reducer(newState, {
          ...action,
          type: "EXPAND",
        });
      }

      if (newState.visibleLeaves[action.source]) {
        return reducer(newState, {
          ...action,
          type: "COLLAPSE",
        });
      }

      return reducer(newState, {
        ...action,
        type: "EXPAND",
      });

    default:
      return newState;
  }
};
