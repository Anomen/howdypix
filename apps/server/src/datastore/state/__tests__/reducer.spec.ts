import { reducer, State } from "../reducer";

describe.skip("state reducer", () => {
  const baseState: State = {
    searches: {},
    searchAlbumRelation: [],
    albums: {
      0: {
        lastUpdatedPhoto: new Date(1),
      },
    },
  };

  it("should update the lastUpdatedPhoto if a more recent photo is saved", () => {
    expect(
      reducer(baseState, {
        type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO",
        albumId: 0,
        data: {
          lastUpdatedPhoto: new Date(2),
        },
      })
    ).toMatchSnapshot();
  });

  it("should NOT update the lastUpdatedPhoto if a LESS recent photo is saved", () => {
    expect(
      reducer(baseState, {
        type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO",
        albumId: 0,
        data: {
          lastUpdatedPhoto: new Date(0),
        },
      })
    ).toMatchSnapshot();
  });

  it("should update if the album lastUpdatedPhoto is null", () => {
    expect(
      reducer(
        { ...baseState, albums: { 0: { lastUpdatedPhoto: null } } },
        {
          type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO",
          albumId: 0,
          data: {
            lastUpdatedPhoto: new Date(0),
          },
        }
      )
    ).toMatchSnapshot();
  });

  it("should not do anything if the album doesn't exist in the state", () => {
    expect(
      reducer(baseState, {
        type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO",
        albumId: 99999,
        data: {
          lastUpdatedPhoto: new Date(0),
        },
      })
    ).toMatchSnapshot();
  });
});
