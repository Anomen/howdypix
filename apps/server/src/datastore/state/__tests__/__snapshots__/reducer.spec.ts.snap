// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state reducer should NOT update the lastUpdatedPhoto if a LESS recent photo is saved 1`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.001Z,
    },
  },
  "searchResults": Object {},
}
`;

exports[`state reducer should NOT update the lastUpdatedPhoto if a LESS recent photo is saved 2`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.001Z,
    },
  },
  "searchResults": Object {},
}
`;

exports[`state reducer should not do anything if the album doesn't exist in the state 1`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.001Z,
    },
  },
  "searchResults": Object {},
}
`;

exports[`state reducer should update if the album lastUpdatedPhoto is null 1`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.000Z,
    },
  },
  "searches": Object {},
}
`;

exports[`state reducer should update the lastUpdatedPhoto if a more recent photo is saved 1`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.002Z,
    },
  },
  "searches": Object {},
}
`;
